### PR TITLE
Ability to update the repository description

### PR DESCRIPTION
**Pull request by adutra**
Note: the update repo functionality is implemented in the old model at adapter level, although, it seems unused. This PR brings the same functionality to the new model.

### CODE REVIEW

**Review by dimas-b — COMMENTED** [review]
Changed LGTM, but it does not look like the new functionality can be accessed. Do you mean to have another PR to expose it somewhere?

**Comment by dimas-b** [comment]
Would it make sense to add a description update command to `quarkus-cli`?

**Comment by adutra** [comment]
> Changes LGTM, but it does not look like the new functionality can be accessed. Do you mean to have another PR to expose it somewhere?

Exactly, the idea is to use this functionality in the importer, in order to "mark" the repository as fully imported. I have a draft already: #7380 .

**Comment by adutra** [comment]
> Would it make sense to add a description update command to `quarkus-cli`?

At this point, that is not required as the only update to the repo description will happen internally inside the importer. We can always add that later though.

File: versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/logic/RepositoryLogicImpl.java [hunk 1]
```diff
@@ -169,14 +184,51 @@ private void addRepositoryDescription(
...
 
+  @Override
+  public RepositoryDescription updateRepositoryDescription(
+      RepositoryDescription repositoryDescription) throws RetryTimeoutException {
+    try {
+      return commitRetry(
+          persist,
+          (p, retryState) -> {
+            try {
+              Reference reference = Objects.requireNonNull(persist.fetchReference(REF_REPO.name()));
```
Comment (dimas-b) [inline versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/logic/RepositoryLogicImpl.java:204]: We should probably fix the javadoc of `REF_REPO`, which currently says `Internal reference with always exactly one commit ...`

File: versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/logic/RepositoryLogicImpl.java [hunk 2]
```diff
@@ -169,14 +184,51 @@ private void addRepositoryDescription(
                   null,
                   "application/json",
                   SHARED_OBJECT_MAPPER.writeValueAsBytes(repoDesc.build()));
-      // can safely ignore the ID returned from storeObj() - it's fine, if the obj already exists
+      // can safely ignore the response from storeObj() - it's fine, if the obj already exists
       persist.storeObj(string);
-      b.addAdds(commitAdd(KEY_REPO_DESCRIPTION, 0, requireNonNull(string.id()), null, null));
+      b.addAdds(
+          commitAdd(KEY_REPO_DESCRIPTION, 0, requireNonNull(string.id()), null, UUID.randomUUID()));
```
Comment (snazy) [inline versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/logic/RepositoryLogicImpl.java:190]: This change will only work for new repositories, but existing repositories do not have a content-ID for the repo-description. The update-logic does not work without a content-ID.

I'd prefer to leave the content-ID as `null`. (There's no need for a CID here)

I know that the repo-config stuff uses a CID, see comment in SLImpl

File: versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/logic/StringLogicImpl.java [hunk 3]
```diff
@@ -68,6 +86,53 @@ public StringObj updateString(StringValue previousValue, String contentType, Str
...
+    StringValue existing = null;
+    if (existingElement != null) {
+      CommitOp op = existingElement.content();
+      if (op.action().exists()) {
+        existingValueId = op.value();
+        existingContentId = op.contentId();
+        existing = existingValueId != null ? fetchString(requireNonNull(existingValueId)) : null;
+      }
+    }
+    if (existingContentId == null) {
```
Comment (snazy) [inline versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/logic/StringLogicImpl.java:116]: This makes updating repo-descriptions of already existing repos impossible (those have no CID for repo-description).

Mabe replacing the `if` with a simple `else` helps here.

Should also have some comment to explain the logic wrt repo-desc + repo-config

**Review by snazy — CHANGES_REQUESTED** [review]

File: versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/logic/RepositoryLogicImpl.java [hunk 4]
```diff
@@ -169,14 +182,60 @@ private void addRepositoryDescription(
...
       b.addAdds(commitAdd(KEY_REPO_DESCRIPTION, 0, requireNonNull(string.id()), null, null));
     } catch (ObjTooLargeException | ObjNotFoundException | IOException e) {
       throw new RuntimeException(e);
     }
   }
 
+  @Override
+  public RepositoryDescription updateRepositoryDescription(RepositoryDescription newDescription)
+      throws RetryTimeoutException {
+    // prevent modification of read-only attributes
```
Comment (adutra) [inline versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/logic/RepositoryLogicImpl.java:196]: This came to my mind after @dimas-b suggestion to introduce an option in quarkus-cli to allow end users to modify the repo description. If we ever do that later, I think we'd need a safety net like this one.

File: versioned/storage/common-tests/src/main/java/org/projectnessie/versioned/storage/commontests/AbstractRepositoryLogicTests.java [hunk 5]
```diff
@@ -246,15 +247,21 @@ public void updateRepositoryDescription() throws RetryTimeoutException {
 
     RepositoryDescription updated =
         RepositoryDescription.builder()
-            .defaultBranchName("main")
             .putProperties("updated", "true")
+            // read-only properties, should be ignored
+            .defaultBranchName("main2")
```
Comment (snazy) [inline versioned/storage/common-tests/src/main/java/org/projectnessie/versioned/storage/commontests/AbstractRepositoryLogicTests.java:252]: Heh, exactly this one might be updated. Let's remove "defaultBranchName" from that check.
I'm not sure whether there'll ever be a command to update arbitrary information - rather per attribute.

File: versioned/storage/common-tests/src/main/java/org/projectnessie/versioned/storage/commontests/AbstractRepositoryLogicTests.java [hunk 6]
```diff
@@ -246,15 +247,21 @@ public void updateRepositoryDescription() throws RetryTimeoutException {
...
             .oldestPossibleCommitTime(Instant.ofEpochSecond(12345))
```
Comment (snazy) [inline versioned/storage/common-tests/src/main/java/org/projectnessie/versioned/storage/commontests/AbstractRepositoryLogicTests.java:253]: This and the below setting are actually read-only attributes.

**Review by snazy — APPROVED** [review]

**Pull request merged**